zebra/router-id.c

```diff
@@ -59,60 +59,107 @@ static struct connected *router_id_find_node(struct list *l,
 	if (!zebra_check_addr(ifc->address))
 		return 1;
 
 	return 0;
 }
 
-void router_id_get(struct prefix *p, vrf_id_t vrf_id)
+static bool router_id_v6_is_any(struct prefix *p)
+{
+	return memcmp(&p->u.prefix6, &in6addr_any, sizeof(struct in6_addr))
+	       == 0;
+}
+
+int router_id_get(afi_t afi, struct prefix *p, vrf_id_t vrf_id)
 {
 	struct listnode *node;
 	struct connected *c;
+	struct in6_addr *addr = NULL;
 	struct zebra_vrf *zvrf = vrf_info_get(vrf_id);
 
-	p->u.prefix4.s_addr = INADDR_ANY;
-	p->family = AF_INET;
-	p->prefixlen = 32;
-
-	if (zvrf->rid_user_assigned.u.prefix4.s_addr != INADDR_ANY)
-		p->u.prefix4.s_addr = zvrf->rid_user_assigned.u.prefix4.s_addr;
-	else if (!list_isempty(zvrf->rid_lo_sorted_list)) {
-		node = listtail(zvrf->rid_lo_sorted_list);
-		c = listgetdata(node);
-		p->u.prefix4.s_addr = c->address->u.prefix4.s_addr;
-	} else if (!list_isempty(zvrf->rid_all_sorted_list)) {
-		node = listtail(zvrf->rid_all_sorted_list);
-		c = listgetdata(node);
-		p->u.prefix4.s_addr = c->address->u.prefix4.s_addr;
+	switch (afi) {
+	case AFI_IP:
+		p->u.prefix4.s_addr = INADDR_ANY;
+		p->family = AF_INET;
+		p->prefixlen = 32;
+		if (zvrf->rid_user_assigned.u.prefix4.s_addr != INADDR_ANY)
+			p->u.prefix4.s_addr =
+				zvrf->rid_user_assigned.u.prefix4.s_addr;
+		else if (!list_isempty(zvrf->rid_lo_sorted_list)) {
+			node = listtail(zvrf->rid_lo_sorted_list);
+			c = listgetdata(node);
+			p->u.prefix4.s_addr = c->address->u.prefix4.s_addr;
+		} else if (!list_isempty(zvrf->rid_all_sorted_list)) {
+			node = listtail(zvrf->rid_all_sorted_list);
+			c = listgetdata(node);
+			p->u.prefix4.s_addr = c->address->u.prefix4.s_addr;
+		}
+		return 0;
+	case AFI_IP6:
+		p->u.prefix6 = in6addr_any;
+		p->family = AF_INET6;
+		p->prefixlen = 128;
+		if (!router_id_v6_is_any(&zvrf->rid6_user_assigned))
+			addr = &zvrf->rid6_user_assigned.u.prefix6;
+		else if (!list_isempty(zvrf->rid6_lo_sorted_list)) {
+			node = listtail(zvrf->rid6_lo_sorted_list);
+			c = listgetdata(node);
+			addr = &c->address->u.prefix6;
+		} else if (!list_isempty(zvrf->rid6_all_sorted_list)) {
+			node = listtail(zvrf->rid6_all_sorted_list);
+			c = listgetdata(node);
+			addr = &c->address->u.prefix6;
+		}
+		if (addr)
+			memcpy(&p->u.prefix6, addr, sizeof(struct in6_addr));
+		return 0;
+	default:
+		return -1;
 	}
 }
 
-static void router_id_set(struct prefix *p, vrf_id_t vrf_id)
+static int router_id_set(afi_t afi, struct prefix *p, vrf_id_t vrf_id)
 {
 	struct prefix p2;
 	struct listnode *node;
 	struct zserv *client;
 	struct zebra_vrf *zvrf;
 
-	if (p->u.prefix4.s_addr == 0) /* unset */
-	{
-		zvrf = vrf_info_lookup(vrf_id);
-		if (!zvrf)
-			return;
-	} else /* set */
-		zvrf = vrf_info_get(vrf_id);
-
-	zvrf->rid_user_assigned.u.prefix4.s_addr = p->u.prefix4.s_addr;
+	switch (afi) {
+	case AFI_IP:
+		if (p->u.prefix4.s_addr == 0) /* unset */
+		{
+			zvrf = vrf_info_lookup(vrf_id);
+			if (!zvrf)
```

Comment:
let's move the zvrf lookup and check for happiness outside of the switch statement it's the same for v4 or v6

Reply:
I don't think it can be moved out as this is looked up only if the address is not set, and how to check it depends on the address type.